src/main/cpp/native_c4replicator.cc

```diff
@@ -234,33 +228,28 @@ static jobject storeContext(JNIEnv* env, jobject jcontext) {
     return gContext;
 }
 
-static void releaseContext(JNIEnv* env, jobject jcontext) {
+// This is *totally* not thread safe
+static void releaseContext(JNIEnv *env, jobject jcontext) {
     if (jcontext == NULL)
         return;
 
     jobject storedContext = NULL;
-    jobject gContext = NULL;
-    if (jcontext != NULL)
-        gContext = env->NewGlobalRef(jcontext);
-
-    if (gContext != NULL) {
-        int i = 0;
-        for (; i < contexts.size(); i++) {
-            jobject c = contexts[i];
-            if (env->IsSameObject(c, gContext)) {
-                storedContext = c;
-                break;
-            }
+    int i = 0;
+    for (; i < contexts.size(); i++) {
+        jobject c = contexts[i];
+        if (env->IsSameObject(c, jcontext)) {
```

Comment:
This PR doesn't address the heart of the problem which is that `c` here is potentially a bad reference because of the non-threadsafe way in which it is released below.  Here is an illustration:

Thread 1 comes in and finds the correct object to store in `storedContext` and makes it as far as `env->DeleteGlobalRef` but not the next `erase` line so right now the collection is still holding a bad ref.

Thread 2 comes in and tries to iterate through each of the objects inside of the collection, but when it gets to the bad one it crashes because a freed ref is passed to `IsSameObject`

Reply:
Nevermind I see now that you've handled this at a higher level.  You should make sure that everything that calls this is also properly synced, or sync here instead of higher.